nixos/modules/services/web-apps/simplesamlphp.nix

```diff
@@ -0,0 +1,140 @@
+  };
+
+  options.services.simplesamlphp = with lib; mkOption {
+    type = types.attrsOf (lib.types.submodule ({ config, ...}: {
+      options = {
+        package = mkPackageOption pkgs "simplesamlphp" { };
+        configureNginx = mkOption {
+          type = types.bool;
+          default = true;
+          description = "Configure nginx as a reverse proxy for SimpleSAMLphp.";
+        };
+        localDomain = mkOption {
+          type = types.str;
+          description = "The domain serving your SimpleSAMLphp instance. This option modifies only /saml";
+        };
+        settings = mkOption {
+          type = lib.types.submodule {
+            freeformType = format.type;
+            options = {
+              baseurlpath = mkOption {
+                type = types.str;
+                example = "https://filesender.example.com/saml/";
+                description = "URL where SimpleSAMLphp can be reached.";
+              };
+            };
+          };
+          default = { };
+          description = ''
+            Configuration options used by SimpleSAMLphp.
+            See [](https://simplesamlphp.org/docs/stable/simplesamlphp-install)
+            for available options.
+          '';
+        };
+
+        authSources = mkOption {
+          type = with types; format.type;
+          default = { };
+          example = { };
+          description = ''
+            Auth sources options used by SimpleSAMLphp.
+          '';
+        };
+
+
+        libDir = mkOption {
+          type = types.str;
+          readOnly = true;
+          description = ''
+            Path to the SimpleSAMLphp library directory.
+
+            ::: {.note}
+            This option is read-only and managed by NixOS.
+            :::
+          '';
+        };
+        configDir = mkOption {
+          type = types.str;
+          readOnly = true;
+          description = ''
+            Path to the SimpleSAMLphp config directory.
+
+            ::: {.note}
+            This option is read-only and managed by NixOS.
+            :::
+          '';
+        };
+      };
+      config = {
+        libDir = "${config.package}/share/php/simplesamlphp/";
+        configDir = "${genSSPConfigForOpts config}";
+      };
+    }));
+    default = {};
+    description = "Instances of SimpleSAMLphp. <name> should refer to valid phpfpm instance.";
+  };
+
+  config = {
+    services.phpfpm.pools = lib.mapAttrs (phpfpmName: opts: let
+      configDir = genSSPConfigForOpts opts;
+    in
+      {
+        phpEnv.SIMPLESAMLPHP_CONFIG_DIR = "${configDir}";
+      }
+    ) cfg;
+
+    services.nginx.virtualHosts = lib.mapAttrs' (phpfpmName: opts: let
+    in
+      lib.nameValuePair opts.localDomain (lib.mkIf opts.configureNginx {
+        locations."^~ /saml/" = {
+          alias = "${opts.package}/share/php/simplesamlphp/www/";
+          extraConfig = ''
+              location ~ ^(?<prefix>/saml)(?<phpfile>.+?\.php)(?<pathinfo>/.*)?$ {
```

Comment:
Why are we not using the location option here?

Reply:
NGINX NixOS module doesn't support nested locations.